proof/refine/ARM/Ipc_R.thy

```diff
@@ -5223,26 +5274,268 @@ lemma replyPush_sch_act_wf:
   apply (wpsimp wp: replyPush_valid_objs' replyPush_sch_act_wf replyPush_valid_queues
                     replyPush_if_live_then_nonz_cap' replyPush_valid_idle'
               simp: valid_pspace'_def)
-  apply (subgoal_tac "invs' s", fastforce)
-  by (simp add: invs'_def valid_state'_def valid_pspace'_def)
+  apply (fastforce simp: invs'_def valid_state'_def valid_pspace'_def
+                         pred_tcb_at'_def obj_at'_def projectKOs
+                  intro: invs_valid_replies')
+  done
 
-lemma setThreadState_Running_invs':
-  "\<lbrace>\<lambda>s. invs' s \<and> tcb_at' t s \<and> ex_nonz_cap_to' t s\<rbrace>
-   setThreadState Running t
+lemma setEndpoint_invs':
+  "\<lbrace>invs' and valid_ep' ep and ex_nonz_cap_to' eptr\<rbrace> setEndpoint eptr ep \<lbrace>\<lambda>_. invs'\<rbrace>"
+  by (wpsimp simp: invs'_def valid_state'_def comp_def)
+
+lemma setSchedContext_valid_ep'[wp]:
+  "setSchedContext param_a param_b \<lbrace>valid_ep' ep\<rbrace>"
+  unfolding valid_ep'_def
+  by (cases ep; wpsimp wp: hoare_vcg_ball_lift)
+
+lemma threadSet_valid_ep'[wp]:
+  "threadSet f tptr \<lbrace>valid_ep' ep\<rbrace>"
+  apply (wpsimp wp: threadSet_wp)
+  apply (fastforce simp: valid_ep'_def obj_at'_def projectKOs objBitsKO_def split: endpoint.splits)
+  done
+
+lemma maybeReturnSc_st_tcb_at'[wp]:
+  "maybeReturnSc nptr tptr \<lbrace>\<lambda>s. P (st_tcb_at' Q t s)\<rbrace>"
+  by (wpsimp wp: hoare_drop_imps threadSet_cap_to' threadSet_pred_tcb_no_state
+           simp: maybeReturnSc_def tcb_cte_cases_def cteCaps_of_def)
+
+crunches maybeReturnSc
+  for sch_act_not[wp]: "sch_act_not t"
+  and sch_act_simple[wp]: "sch_act_simple"
+  and ep_at'[wp]: "\<lambda>s. P (ep_at' eptr s)"
+  and reply_at'[wp]: "\<lambda>s. P (reply_at' rptr s)"
+  and valid_ep'[wp]: "valid_ep' ep"
+  and valid_tcb_state'[wp]: "valid_tcb_state' tcb"
+  and valid_bound_reply'[wp]: "valid_bound_reply' reply"
+  (wp: crunch_wps hoare_drop_imps simp: crunch_simps)
+
+crunches maybeReturnSc, doIPCTransfer
+  for replyTCB_obj_at'[wp]: "\<lambda>s. P (obj_at' (\<lambda>reply. P' (replyTCB reply)) t s)"
+  (wp: crunch_wps constOnFailure_wp simp: crunch_simps)
+
+lemma blockedCancelIPC_st_tcb_at:
+  "\<lbrace>\<lambda>s. (t = t' \<longrightarrow> Q (P Inactive)) \<and> (t \<noteq> t' \<longrightarrow> Q (st_tcb_at' P t s))\<rbrace>
+   blockedCancelIPC st t' rptr
+   \<lbrace>\<lambda>_ s. Q (st_tcb_at' P t s)\<rbrace>"
+  unfolding blockedCancelIPC_def getBlockingObject_def
+  by (wpsimp wp: sts_st_tcb_at'_cases_strong replyUnlink_st_tcb_at' getEndpoint_wp hoare_vcg_imp_lift')
+
+lemma replyRemoveTCB_st_tcb_at'_cases:
+  "\<lbrace>\<lambda>s. (t = t' \<longrightarrow> Q (P Inactive)) \<and> (t \<noteq> t' \<longrightarrow> Q (st_tcb_at' P t s))\<rbrace>
+   replyRemoveTCB t'
+   \<lbrace>\<lambda>_ s. Q (st_tcb_at' P t s)\<rbrace>"
+  unfolding replyRemoveTCB_def cleanReply_def
+  by (wpsimp wp: replyUnlink_st_tcb_at' gts_wp' hoare_vcg_imp_lift')
```

Comment:
Something very similar exists in Reply_R. Can you please generalise that one rather than adding a new one here?

Reply:
I think this was my intention, but I forgot to make a note